Makefile

```diff
@@ -103,31 +104,32 @@ lib-token: ${CACHEDIR}/.libraries_token.txt
 
 .PHONY: fetch-kernel
 fetch-kernel:
-	rm -rf kernel
-	$(MAKE) kernel/boot/vmlinuz
+	rm -rf kernel/$(ARCH)
+	$(MAKE) kernel/$(ARCH)/vmlinuz
 
-kernel/APKINDEX.tar.gz:
+kernel/%/APKINDEX.tar.gz:
 	@$(call authget,apk.cgr.dev,$@,$(QEMU_KERNEL_REPO)/$(ARCH)/APKINDEX.tar.gz)
 
-kernel/APKINDEX: kernel/APKINDEX.tar.gz
-	tar -x -C kernel -f $< $(notdir $@)
+kernel/%/APKINDEX: kernel/%/APKINDEX.tar.gz
+	tar -x -C kernel --to-stdout -f $< APKINDEX > $@.tmp.$$$$ && mv $@.tmp.$$$$ $@
 	touch $@
 
-kernel/chosen: kernel/APKINDEX
+kernel/%/chosen: kernel/%/APKINDEX
 	# Extract lines with 'P:linux' and the following line that contains the version
 	# This approach is compatible with both GNU and BSD sed
-	awk '/^P:linux$$/ {print; getline; print}' $< > kernel/available
-	grep '^V:' kernel/available | sed 's/V://' | \
+	awk '/^P:linux$$/ {print; getline; print}' $< > kernel/$*/available
+	grep '^V:' kernel/$*/available | sed 's/V://' | \
 	  sort -V | tail -n1 > $@.tmp
 	# Sanity check that this looks like an apk version
 	grep -E '^([0-9]+\.)+[0-9]+-r[0-9]+$$' $@.tmp
 	mv $@.tmp $@
 
-kernel/linux.apk: kernel/chosen
-	@$(call authget,apk.cgr.dev,$@,$(QEMU_KERNEL_REPO)/$(ARCH)/linux-$(shell cat kernel/chosen).apk)
+kernel/%/linux.apk: kernel/%/chosen
+	@$(call authget,apk.cgr.dev,$@,$(QEMU_KERNEL_REPO)/$*/linux-$(shell cat kernel/$*/chosen).apk)
 
-kernel/boot/vmlinuz: kernel/linux.apk
-	tar -x -C kernel -f $< boot/ 2> /dev/null
+kernel/%/vmlinuz: kernel/%/linux.apk
+	tmpd=kernel/.$$$$ && mkdir -p $$tmpd $(dir $@) && tar -x -C $$tmpd -f $< boot/ 2> /dev/null && \
+		mv $$tmpd/boot/* $(dir $@) && rc=$$?; rm -Rf $$tmpd; exit $$rc
```

Comment:
maybe we could avoid the tmpd with `tar --strip-components=1` ?

```
dann-frazier@dann-frazier-laptop:~/git/os$ find kernel/x86_64/
kernel/x86_64/
kernel/x86_64/linux.apk
dann-frazier@dann-frazier-laptop:~/git/os$ tar -x -C kernel/x86_64/ -f kernel/x86_64/linux.apk boot/vmlinuz --strip-components=1
tar: Ignoring unknown extended header keyword 'APK-TOOLS.checksum.SHA1'
tar: Ignoring unknown extended header keyword 'APK-TOOLS.checksum.SHA1'
tar: Ignoring unknown extended header keyword 'APK-TOOLS.checksum.SHA1'
tar: Ignoring unknown extended header keyword 'APK-TOOLS.checksum.SHA1'
dann-frazier@dann-frazier-laptop:~/git/os$ find kernel/x86_64/
kernel/x86_64/
kernel/x86_64/vmlinuz
kernel/x86_64/linux.apk
```

Reply:
i dont have strong feelings either way. its a tmpdir and it gets cleaned up.

Reply:
I just added another change which i think actually justifies the tmpdir. Now we verify that we actually downloaded a vmlinuz rather than just touch-updated a potentially old one.